### PR TITLE
`cast_possible_truncation` Suggest TryFrom when truncation possible

### PR DESCRIPTION
**Pull request by xFrednet**
This fixes the last issues from https://github.com/rust-lang/rust-clippy/pull/9664 as the author seems to be inactive. The PR author was sadly kept during the rebase, due to the conflict resolution.

IDK if it's worth it do to a full review, I only added the last commit, everything else remained the same, besides a rebase.

---

changelog: Sugg: [`cast_possible_truncation`]: Now suggests using `try_from` or allowing the lint
[#10038](https://github.com/rust-lang/rust-clippy/pull/10038)

closes: https://github.com/rust-lang/rust-clippy/issues/9231

### CODE REVIEW

File: clippy_lints/src/casts/cast_possible_truncation.rs [hunk 1]
```diff
@@ -153,5 +155,19 @@ pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, cast_expr: &Expr<'_>,
         _ => return,
     };
 
-    span_lint(cx, CAST_POSSIBLE_TRUNCATION, expr.span, &msg);
+    let snippet = snippet(cx, expr.span, "..");
+    let name_of_cast_from = snippet.split(" as").next().unwrap_or("..");
```
Comment (Jarcho) [inline clippy_lints/src/casts/cast_possible_truncation.rs:159]: This doesn't work with chained casts. e.g. `x as u32 as u16`.

File: clippy_lints/src/casts/cast_possible_truncation.rs [hunk 2]
```diff
@@ -153,5 +155,19 @@ pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, cast_expr: &Expr<'_>,
...
+    let suggestion = format!("{cast_to}::try_from({name_of_cast_from})");
```
Comment (Jarcho) [inline clippy_lints/src/casts/cast_possible_truncation.rs:160]: Using `cast_to` in the suggestion will lose any type aliases. It would be better to take a snippet.

**Review by Jarcho — COMMENTED** [review]
Just a couple of edge cases noted. Otherwise LGTM.

**Comment by xFrednet** [comment]
Good points! I've rebased, since it has been over a month. All changes are in the last commit, as well as a test for the `as u32 as u16` expression.

Thank you for your patience, when it came to the update :)

**Comment by Jarcho** [comment]
Thank you. @bors r+

**Comment by RalfJung** [comment]
I am a bit surprised that this was marked as closing https://github.com/rust-lang/rust-clippy/issues/9231; at least I don't see a test for usize-to-u64 casts here.
